filter-kalman/kalman.cpp

```diff
@@ -9,73 +9,212 @@
 #include <QDebug>
 #include <cmath>
 
-constexpr double settings::mult_noise_stddev;
+
+void KalmanFilter::init()
+{
+    static constexpr int NS = NUM_STATE_DOF;
+    static constexpr int NZ = NUM_MEASUREMENT_DOF;
+    // allocate and initialize matrices
+    measurement_noise_cov = Matrix::Zero(NZ, NZ);
+    process_noise_cov = Matrix::Zero(NS, NS);
+    kalman_gain = Matrix::Zero(NS, NZ);
+    measurement_matrix = Matrix::Zero(NZ, NS);
+    state_cov = Matrix::Zero(NS, NS);
+    state_cov_prior = Matrix::Zero(NS, NS);
+    transition_matrix = Matrix::Zero(NS, NS);
+    // initialize state variables
+    state = StateVector::Zero();
+    state_prior = StateVector::Zero();
+    innovation = PoseVector::Zero();
+}
+
+
+void KalmanFilter::time_update()
+{
+    state_prior     = transition_matrix * state;
+    state_cov_prior = transition_matrix * state_cov * transition_matrix.transpose() + process_noise_cov;
+}
+
+
+void KalmanFilter::measurement_update(const PoseVector &measurement)
+{
+    Matrix tmp     = measurement_matrix * state_cov_prior * measurement_matrix.transpose() + measurement_noise_cov;
+    Matrix tmp_inv = tmp.inverse();
+    kalman_gain = state_cov_prior * measurement_matrix.transpose() * tmp_inv;
+    innovation = measurement - measurement_matrix * state_prior;
+    state     = state_prior + kalman_gain * innovation;
+    state_cov = state_cov_prior - kalman_gain * measurement_matrix * state_cov_prior;
+}
+
+
+
+void KalmanProcessNoiseScaler::init()
+{
+    base_cov = Matrix::Zero(NUM_STATE_DOF, NUM_STATE_DOF);
+    innovation_cov_estimate = Matrix::Zero(NUM_MEASUREMENT_DOF, NUM_MEASUREMENT_DOF);
+}
+
+
+/* Uses 
+    innovation, measurement_matrix, measurement_noise_cov, and state_cov_prior
+   found in KalmanFilter. It sets 
+    process_noise_cov
+*/
+void KalmanProcessNoiseScaler::update(KalmanFilter &kf, double dt)
+{
+    Matrix ddT = kf.innovation * kf.innovation.transpose();
+    double f = dt / (dt + settings::adaptivity_window_length);
+    innovation_cov_estimate =
+        f * ddT + (1. - f) * innovation_cov_estimate;
+
+    double T1 = (innovation_cov_estimate - kf.measurement_noise_cov).trace();
+    double T2 = (kf.measurement_matrix * kf.state_cov_prior * kf.measurement_matrix.transpose()).trace();
+    double alpha = 0.001;
+    if (T2 > 0. && T1 > 0.)
+    {
+        alpha = T1 / T2;
+        alpha = std::sqrt(alpha);
+        alpha = std::min(1000., std::max(0.001, alpha));
+    }
+    kf.process_noise_cov = alpha * base_cov;
+    //qDebug() << "alpha = " << alpha;
+}
+
+
+PoseVector DeadzoneFilter::filter(const PoseVector &input)
+{
+    PoseVector out;
+    for (int i = 0; i < input.rows(); ++i)
+    {
+        const double dz = dz_size[i];
```

Comment:
Maybe add some small epsilon like `1e-8` to the deadzone? It looks easy for it to be zero. We flush denormals btw.